src/middleware/routeHandler.ts

```diff
@@ -0,0 +1,250 @@
+  if (result.isErr()) {
+    next(result.error)
+    return false
+  }
+  const isGitLocked = result.value
+  if (isGitLocked) {
+    logger.error(`Failed to lock repo ${repoName}: git file system in use.`)
+    next(
+      new LockedError(
+        `Someone else is currently modifying repo ${repoName}. Please try again later.`
+      )
+    )
+    return false
+  }
+  return true
+}
+
+// Used when there are no write API calls to the repo on GitHub
+export const attachReadRouteHandlerWrapper = (routeHandler: any) => async (
+  req: any,
+  res: any,
+  next: any
+) => {
+  routeHandler(req, res).catch((err: any) => {
+    next(err)
+  })
+}
+
+// Used when there are write API calls to the repo on GitHub
+export const attachWriteRouteHandlerWrapper = (routeHandler: any) => async (
+  req: any,
+  res: any,
+  next: any
+) => {
+  const { siteName } = req.params
+  const { growthbook } = req
+
+  let isGitAvailable = true
+
+  // only check git file lock if the repo is ggs enabled
+  if (growthbook?.getFeatureValue(FEATURE_FLAGS.IS_GGS_ENABLED, false)) {
+    isGitAvailable = await handleGitFileLock(siteName, next)
+  }
+
+  if (!isGitAvailable) return
+
+  try {
+    await lock(siteName)
+  } catch (err) {
+    next(err)
+    return
+  }
+
+  await routeHandler(req, res, next).catch(async (err: any) => {
+    await unlock(siteName)
+    next(err)
+  })
+
+  try {
+    await unlock(siteName)
+  } catch (err) {
+    next(err)
+  }
+}
+
+export const attachRollbackRouteHandlerWrapper = (routeHandler: any) => async (
+  req: any,
+  res: any,
+  next: any
+) => {
+  const { userSessionData } = res.locals
+  const { siteName } = req.params
+
+  const { accessToken } = userSessionData
+  const { growthbook } = req
+
+  const shouldUseGitFileSystem = !!growthbook?.getFeatureValue(
+    FEATURE_FLAGS.IS_GGS_ENABLED,
+    false
+  )
+
+  const isGitAvailable = await handleGitFileLock(siteName, next)
+  if (!isGitAvailable) return
+  try {
+    await lock(siteName)
+  } catch (err) {
+    next(err)
+    return
+  }
+
+  let originalStagingCommitSha: any
+  let originalStagingLiteCommitSha: any
+
+  if (shouldUseGitFileSystem) {
+    const results = await Promise.all([
+      gitFileSystemService.getLatestCommitOfBranch(siteName, STAGING_BRANCH),
+      gitFileSystemService.getLatestCommitOfBranch(
+        siteName,
+        STAGING_LITE_BRANCH
+      ),
+    ])
+    const [stagingResult, stagingLiteResult] = results
+
+    if (stagingResult.isErr() || stagingLiteResult.isErr()) {
+      await unlock(siteName)
+      if (stagingResult.isErr()) next(stagingResult.error)
+      if (stagingLiteResult.isErr()) next(stagingLiteResult.error)
+      return
+    }
+    originalStagingCommitSha = stagingResult.value.sha
+    originalStagingLiteCommitSha = stagingLiteResult.value.sha
+    if (!originalStagingCommitSha || !originalStagingLiteCommitSha) {
+      await unlock(siteName)
+      if (stagingResult.isErr()) next(stagingResult.error)
```

Comment:
this should never hit because we will do an early return previously. maybe we can pass a string or something of the same shape if `next` does work on the `error` itself.

Reply:
[79cdf6a](https://github.com/isomerpages/isomercms-backend/pull/1044/commits/79cdf6a57e0888d1011884f852ef834e562605ca)